karapace/schema_backup.py

```diff
@@ -310,10 +345,16 @@ def restore_backup(self) -> None:
             LOG.info("Starting backup restore for topic: %r", self.topic_name)
 
             with open(self.backup_location, encoding="utf8") as fp:
-                if _check_backup_file_version(fp) == BackupVersion.V2:
-                    self._restore_backup_version_2(producer, fp)
+                version = _read_backup_file_version(fp)
```

Comment:
This could return the correct backend, no need to have `if-elif-else`:
```suggestion

class BackupBackend:
    def items_from_file(self, fp) ->Generator[tuple[str, str], None, None]:
        ...
        
    def serialize_record(self, key_bytes: bytes | None, value_bytes: bytes | None) -> str:
        ...

class V2BackupBackend(BackupBackend):
    ...
---------------

                backend = _select_backup_backend(fp)
                items = backend.items_from_file(fp)
```

The API would also be formalized.

Reply:
Hmm, I wrote a long response here on Friday, it seems to have been lost, maybe I forgot to post it.

---

The reason I did not opt to introduce an interface here is because the three versions will be sufficiently different for us to have any valuable gains from a shared interface.

- V1 is only readable, so the only thing it has in common with V2 is the signature of the reader: `(fp: IO) -> Iterator[tuple[str, str]]`.
- V1 and V2 are dealing with `str`, while I'm very much hoping that V3 won't have to make any UTF-8 interpretations at all, i.e. exclusively deal with `bytes`.
- V1 and V2 takes `key` + `value` for serialization. In V3 we need a much richer object to be able to pass all the required metadata, so a shared interface doesn't fit well here either. 

Since the current state of this PR mostly moves things around, I'm hoping we can find the value in its current scope sufficient to move forward, and then consider whether we additionally need further refactoring.

Reply:
I can take another round at formalizing the interface.